crates/re_viewer/src/app.rs

```diff
@@ -1751,53 +1735,49 @@ fn save_database_to_file(
     path: std::path::PathBuf,
     time_selection: Option<(re_data_store::Timeline, TimeRangeF)>,
 ) -> impl FnOnce() -> anyhow::Result<std::path::PathBuf> {
-    use re_log_types::{EntityPathOpMsg, TimeInt};
-
-    let msgs = match time_selection {
-        // Fast path: no query, just dump everything.
-        None => log_db
-            .chronological_log_messages()
-            .cloned()
-            .collect::<Vec<_>>(),
-
-        // Query path: time to filter!
-        Some((timeline, range)) => {
-            use std::ops::RangeInclusive;
-            let range: RangeInclusive<TimeInt> = range.min.floor()..=range.max.ceil();
-            log_db
-                .chronological_log_messages()
-                .filter(|msg| {
-                    match msg {
-                        LogMsg::BeginRecordingMsg(_) | LogMsg::Goodbye(_) => {
-                            true // timeless
-                        }
-                        LogMsg::EntityPathOpMsg(EntityPathOpMsg { time_point, .. }) => {
-                            time_point.is_timeless() || {
-                                let is_within_range = time_point
-                                    .get(&timeline)
-                                    .map_or(false, |t| range.contains(t));
-                                is_within_range
-                            }
-                        }
-                        LogMsg::ArrowMsg(_) => {
-                            // TODO(john)
-                            false
-                        }
-                    }
-                })
-                .cloned()
-                .collect::<Vec<_>>()
-        }
-    };
+    use re_arrow_store::TimeRange;
+
+    crate::profile_scope!("dump_messages");
+
+    let begin_rec_msg = log_db
+        .recording_msg()
+        .map(|msg| LogMsg::BeginRecordingMsg(msg.clone()));
+
+    let ent_op_msgs = log_db
+        .iter_entity_op_msgs()
+        .map(|msg| LogMsg::EntityPathOpMsg(msg.clone()))
+        .collect_vec();
+
+    let time_filter = time_selection.map(|(timeline, range)| {
+        (
+            timeline,
+            TimeRange::new(range.min.floor(), range.max.ceil()),
+        )
+    });
+    let data_msgs: Result<Vec<_>, _> = log_db
+        .entity_db
+        .data_store
+        .to_data_tables(time_filter)
+        .map(|table| table.to_arrow_msg().map(LogMsg::ArrowMsg))
+        .collect();
+
+    let msgs = data_msgs.map(|data_msgs| {
+        std::iter::once(begin_rec_msg)
+            .flatten()
+            .chain(ent_op_msgs)
+            .chain(data_msgs)
+    });
 
     move || {
         crate::profile_scope!("save_to_file");
 
+        let msgs = msgs.with_context(|| format!("Failed to create file at {path:?}"))?;
+
```

Comment:
bad error message, and weirdly late handling of the error

```suggestion
    let data_msgs = data_msgs.with_context("Failed to export to data tables")?;

    let msgs =
        std::iter::once(begin_rec_msg)
            .flatten()
            .chain(ent_op_msgs)
            .chain(data_msgs);

    move || {
        crate::profile_scope!("save_to_file");

```